pkg/crosscluster/logical/batch_handler_test.go

```diff
@@ -0,0 +1,391 @@
+	indexEntries, err := rowenc.EncodePrimaryIndex(
+		b.codec,
+		b.tableDesc,
+		b.tableDesc.GetPrimaryIndex(),
+		b.colMap,
+		row,
+		false,
+	)
+	require.NoError(b.t, err)
+	require.Len(b.t, indexEntries, 1)
+	kv := roachpb.KeyValue{
+		Key:   indexEntries[0].Key,
+		Value: indexEntries[0].Value,
+	}
+	kv.Value.Timestamp = timestamp
+	kv.Value.InitChecksum(kv.Key)
+	return kv
+}
+
+// insertEvent creates an insert event for the given row at the specified timestamp.
+func (b *EventBuilder) insertEvent(time hlc.Timestamp, row tree.Datums) streampb.StreamEvent_KV {
+	event := streampb.StreamEvent_KV{
+		KeyValue: b.encodeRow(time, row),
+	}
+	return event
+}
+
+// updateEvent creates an update event for the given row and previous values at the specified timestamp.
+func (b *EventBuilder) updateEvent(
+	time hlc.Timestamp, row tree.Datums, prevValue tree.Datums,
+) streampb.StreamEvent_KV {
+	kv := b.encodeRow(time, row)
+	kvPrev := b.encodeRow(time, prevValue)
+	return streampb.StreamEvent_KV{
+		KeyValue:  kv,
+		PrevValue: kvPrev.Value,
+	}
+}
+
+// deleteEvent creates a delete event for the given row at the specified timestamp.
+func (b *EventBuilder) deleteEvent(
+	time hlc.Timestamp, prevValue tree.Datums,
+) streampb.StreamEvent_KV {
+	kv := b.encodeRow(time, prevValue)
+	return streampb.StreamEvent_KV{
+		KeyValue: roachpb.KeyValue{
+			Key:   kv.Key,
+			Value: roachpb.Value{Timestamp: time},
+		},
+		PrevValue: kv.Value,
+	}
+}
+
+// newBatchHandler creates a new batch handler for testing.
+func newBatchHandler(
+	t *testing.T, s serverutils.ApplicationLayerInterface, tableName string,
+) (*kvRowProcessor, catalog.TableDescriptor) {
+	ctx := context.Background()
+	desc := cdctest.GetHydratedTableDescriptor(t, s.ExecutorConfig(), tree.Name(tableName))
+	sd := sql.NewInternalSessionData(ctx, s.ClusterSettings(), "" /* opName */)
+	handler, err := newKVRowProcessor(
+		ctx,
+		&execinfra.ServerConfig{
+			DB:           s.InternalDB().(descs.DB),
+			LeaseManager: s.LeaseManager(),
+			Settings:     s.ClusterSettings(),
+		},
+		&eval.Context{
+			Codec:            s.Codec(),
+			Settings:         s.ClusterSettings(),
+			SessionDataStack: sessiondata.NewStack(sd),
+		},
+		execinfrapb.LogicalReplicationWriterSpec{},
+		map[descpb.ID]sqlProcessorTableConfig{
+			desc.GetID(): {
+				srcDesc: desc,
+			},
+		},
+	)
+	require.NoError(t, err)
+	return handler, desc
+}
+
+// addAndRemoveColumn is a metamorphic test option that randomly decides whether
+// to add and remove a column to trigger a rebuild of the primary key.
+var addAndRemoveColumn = metamorphic.ConstantWithTestBool("batch-handler-exhaustive-rebuild-primary-index", false)
+
+// uniqueConstraint is a metamorphic test option that randomly decides whether
+// to add a unique constraint to the table.
+var uniqueConstraint = metamorphic.ConstantWithTestBool("batch-handler-exhaustive-unique-constraint", false)
```

Comment:
the presence of a unique constraint should not affect results, as we don't have concurrent writers to the key space, right?